[![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Logo.jpg)](https://archive.stsci.edu/)

## Table of Contents
- [Overview](#overview)
- [MAST Portal](#mast-portal)
    - [Searching for Data](#searching-for-data)
    - [Browsing Data](#browsing-data)
    - [Browsing Tools](#browsing-tools)
    - [Data Retrieval](#data-retrieval)
    - [Video Tutorials](#video-tutorials)
- [MAST API via astroquery](#mast-api-via-astroquery)
    - [Observational Queries](#observational-queries)
    - [Downloading Data Products](#downloading-data-products)
    - [Catalog Queries](#catalog-queries)
- [Exoplanet Atmosphere Observability Table](#exoplanet-atmosphere-observability-table)
- [Exercises](#exercises)
    - [1. Download TESS Light Curve Data](#1-download-tess-light-curve-data)
    - [2. Previewing Hubble Data for Transmission Spectroscopy](#2-previewing-hubble-data-for-transmission-spectroscopy)
    - [3. Using the NASA Exoplanet Archive with MAST to Systematically Access Exoplanet Host Data](#3-using-the-nasa-exoplanet-archive-with-mast-to-systematically-access-exoplanet-host-data)

# Overview
The Barbara A. Mikulski Archive for Space Telescopes (MAST) is the primary data repository for all space telescope missions, including Hubble, JWST, TESS, and Kepler. Most exoplanet discoveries and analysis are performed using these instruments, so being able to navigate, query, and download data products is imperative. MAST is a **vast** network with countless tools, many of which are out of scope for this curriculum. Consequently, this module covers the following tools:
- [MAST Portal](https://mast.stsci.edu/portal/Mashup/Clients/Mast/Portal.html)
- [MAST API via astroquery](https://astroquery.readthedocs.io/en/latest/mast/mast.html)
- [Exoplanet Atmosphere Observability Table](https://catalogs.mast.stsci.edu/eaot/)

Resources that are not discussed, but that may be useful to some users, include:
- [TESScut (TESS Full Frame Image Cutout Tool)](https://mast.stsci.edu/tesscut/)
- [MAST Search (Hubble and JWST form-based query tool)](https://mast.stsci.edu/search/ui/#/)
- [MAST API via HTTPS Requests](https://mast.stsci.edu/api/v0/)

# [MAST Portal](https://mast.stsci.edu/portal/Mashup/Clients/Mast/Portal.html)
The MAST Portal is the web-based GUI where data can be easily accessed from. This section will go over how to construct a basic query and will offer pointers to tutorials and user guides for more advanced queries and tools.

## Searching for Data

### [Basic Search](https://outerspace.stsci.edu/display/MASTDOCS/Basic+Search)
This comes in two steps:
- Select a collection to query
- Submitting a valid search criteria

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_menu_selectCollection.png)

In general, 'MAST Observations' is a safe default. A [table](https://outerspace.stsci.edu/display/MASTDOCS/Basic+Search) is available that gives information on the other collections.

There are two main ways to query an object: by **name** or **coordinates**. 

#### Search by Name
Querying by object is straightforward, and MAST is able to recognize a variety of different inputs for the same target. For example, WASP-100 also goes by TIC 38846515, 2MASS J04355033-6401373, etc., and will result in the same query. Here is a table outlining common paradigms:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Allowed.PNG)

#### Search by Coordinate
RA and DEC can be specified along with a search radius to query a region of the sky. The format of this search is:

`RA DEC r=[radius][unit]`

Where `unit` can be "d" (degrees), "min" (arcminutes), or "s" (arc-seconds). RA and DEC units can be in sexadecimal, decimal degrees or galactic coordinates. Below outlines the acceptable formats:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Allowed_Coord.PNG)

### [List Search](https://outerspace.stsci.edu/display/MASTDOCS/Search+a+List+of+Targets)
A list of targets can also be passed as a search.

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Target_List.PNG)

The minimum required information, in distinct columns, are **Target** and **RA + DEC**, following the same paradigms as the basic search. 

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_List_Example.PNG)

More details and caveats to this process can be found [here](https://outerspace.stsci.edu/display/MASTDOCS/Search+a+List+of+Targets).

## Browsing Data
After searching for your target(s), a list of available data products appear with a variety of information. These can be continually filtered, interacted with, and previewed, all within the portal interface. This section briefly discusses the basics of the interface, with pointers to additional information for more advanced searches.

### Results Grid
Let's say we search the system WASP-10 using a basic search. The results grid will look like:
![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Table.PNG)

The columns provide information such as filters, instrument, project, etc. The 'Edit Columns' in the upper left allow you to pick and choose which to display, and are explained [here](https://outerspace.stsci.edu/display/MASTDOCS/Search+Results+Grid). 

There are a number of icons that allow you to perform certain actions, if applicable, summarized in the table below:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Actions.PNG)

For our target, we can easily download the TESS time-series data, which would be the second row in the list displayed above (the first is an FFI, or Full Frame Image) by clicking on the "Save" icon.

### Filters
To refine the output of the results grid, we can apply a variety of different filters. Some example filters are shown here:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Filters.PNG)

These are all explanatory, so feel free to explore and apply filters as you see fit.

## Browsing Tools

### [AstroView](https://outerspace.stsci.edu/display/MASTDOCS/AstroView)

When searching for data, MAST will give a close-up sky view of the target in a box called `AstroView`, shown below:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Astroview.PNG)

Each orange outline/dot represents an associated data product with that part of the sky, which can be clicked on to highlight the data set in the results grid. **These regions are not necessarily tied to your searched target, which is indicated by the red crosshair. Be sure to check the target associated with the highlighted product before downloading.**

### Actions and Visualization
Some data allow you to perform actions or visualize data in-broswer, which can be useful to see before you download the data product. These are summarized in the table below:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Visual.PNG)

The details of these tools and how to use the interfaces can be found [here](https://outerspace.stsci.edu/display/MASTDOCS/Data+Browsing+Tools).


### Downloading the Results Grid

In some contexts, downloading the current results grid to an external file can be necessary. This can be done by clicking ![Logo ><](../Graphics/Data_Repositories/MAST_Download.PNG), and filling out the `Export Table` pop-up, shown below:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_GridExport.PNG)

## Data Retrieval

### One-Click Download
As described above, data can be downloaded in one click using the ![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_Save.png) icon next to the associated data product, which will download the **minimum recommended data products** for that observation. The file will save as a `.zip` file, which will have to be unpacked. That's it!

### Download Basket

If more than one data set is desired, they can be selected in the left most column in the results grid, shown here:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_GridSelect.PNG)

These can then be added to the download basket using ![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_DownloadBasket.PNG). This will bring up the download manager window:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_DownloadManage.PNG)

Where specific files can be selected, filtered, or ignored, before being downloaded. See [here](https://outerspace.stsci.edu/display/MASTDOCS/Download+Basket) for more information on file types, filters, details and actions

## Video Tutorials

[Video demos](https://outerspace.stsci.edu/display/MASTDOCS/Demos+and+Tutorials) are available online for a comprehensive dive into the MAST portal!

# [MAST API via astroquery](https://astroquery.readthedocs.io/en/latest/mast/mast.html)

The MAST portal can be effective when dealing with a few targets, but struggles when target lists reach the **hundreds** or ***thousands***. For this, the Python module `astroquery` can retrieve and download associated data products from MAST programmatically. For example, downloading all TESS time-series photometry for all known TESS Objects of Interest can be reduced down to a list of TIC IDs and a simple `for` loop. The query types addressed in this module are:

- [Observational Queries](https://astroquery.readthedocs.io/en/latest/mast/mast_obsquery.html)
- [Catalog Queries](https://astroquery.readthedocs.io/en/latest/mast/mast_catalog.html)

The queries not addressed here, but may be useful, are:

- [Mission Searches (JWST, HST metadata)](https://astroquery.readthedocs.io/en/latest/mast/mast_missions.html)
- [Image Cutouts (e.g. TESScut)](https://astroquery.readthedocs.io/en/latest/mast/mast_cut.html)
- [MAST Queries](https://astroquery.readthedocs.io/en/latest/mast/mast_mastquery.html)

This section goes through the basic functions of `astroquery.mast`, with a few illustrative examples. To start, we import the appropriate modules:

In [15]:
from astroquery.mast import Observations, Catalogs

## [Observational Queries](https://astroquery.readthedocs.io/en/latest/api/astroquery.mast.ObservationsClass.html#astroquery.mast.ObservationsClass)
### Generating Observation Tables
The `Observations` class input is similar to the MAST Portal's filtering. Queries can be made by **position**, **target name**, or other **criteria** (such as catalog ID number). These queries output an `astropy` table of observations, which can then be downloaded. Here are examples of typical workflows.

In [16]:
##Query the exoplanet HAT-P-7b
query_table = Observations.query_object('HAT-P-7b')
print(query_table)

intentType obs_collection provenance_name ...   obsid        distance     
---------- -------------- --------------- ... --------- ------------------
   science           TESS            SPOC ...  27463641                0.0
   science           TESS            SPOC ...  27507606                0.0
   science           TESS            SPOC ...  62431376                0.0
   science           TESS            SPOC ...  62870786                0.0
   science           TESS            SPOC ...  92616930                0.0
   science           TESS            SPOC ...  95133384                0.0
   science           TESS            SPOC ... 212510190                0.0
   science           TESS            SPOC ... 213002720                0.0
   science           TESS            SPOC ... 230167024                0.0
   science           TESS            SPOC ... 232881378                0.0
       ...            ...             ... ...       ...                ...
   science            HLA

That took some time! Most exoplanet hosts will have a comparable number of observations. While there are some uses for a broad search, let's look at some applications for adding some parameters. 

Filtering can be done by constraining the content of each column. You can find a complete list of all the columns and their descriptions [here](https://mast.stsci.edu/api/v0/_c_a_o_mfields.html).

The `query_object` only allows you to add a "radius of search" parameter object, so we will use the `query_criteria` method. 

Let's look for light curves of HAT-P-7b taken by the TESS telescope. This allows us to narrow down our search significantly!

In [17]:
query_table = Observations.query_criteria(objectname="HAT-P-7b", 
                                              obs_collection="TESS",
                                              dataproduct_type=["timeseries"])
# indenting like this doesn't affect the code output, but is just a way to try and make it more readable/organized

print(len(query_table))
print(query_table[:10])

45
intentType obs_collection provenance_name ...   objID1       distance     
---------- -------------- --------------- ... --------- ------------------
   science           TESS            SPOC ...  70526761 287.98135262165573
   science           TESS            SPOC ...  70531322                0.0
   science           TESS            SPOC ...  70544780  638.6413483917396
   science           TESS            SPOC ... 116683518                0.0
   science           TESS            SPOC ... 116846050                0.0
   science           TESS            SPOC ... 117053213                0.0
   science           TESS            SPOC ... 117053215 118.92799122073585
   science           TESS            SPOC ... 120248572                0.0
   science           TESS            SPOC ... 120248575 118.92799122073585
   science           TESS            SPOC ... 120248644 118.92799122073585


 <div class="alert alert-block alert-info">
 
**INFO**: The **query_criteria** method is extremely robust, and can suit most search applications extremely efficiently. The different parameters to refine your search can be found [here](https://mast.stsci.edu/api/v0/_c_a_o_mfields.html)

Lastly, we can also query by position by using the `query_region` method, which takes in an RA, DEC, and search radius, and outputs all data in the region.

In [18]:
##The syntax is "RA DEC", in any commonly accepted RA DEC format as outlined in "Search by Coordinate" section
query_table = Observations.query_region("100.12 -20.223")
print(query_table)

intentType obs_collection  provenance_name  ...   obsid        distance    
---------- -------------- ----------------- ... --------- -----------------
   science           TESS              SPOC ...  62280343               0.0
   science           TESS              SPOC ...  62324728               0.0
   science           TESS              SPOC ...  28287315               0.0
   science           TESS              SPOC ...  28220318   642.16680527556
   science            PS1               3PI ...   1850796               0.0
   science            PS1               3PI ...   1850797               0.0
   science            PS1               3PI ...   1850798               0.0
   science            PS1               3PI ...   1850799               0.0
   science            PS1               3PI ...   1850800               0.0
   science            PS1               3PI ...   1850791 503.6552188317507
       ...            ...               ... ...       ...               ...
   science  

### Observation Counts
Each method above has a corresponding 'count' method (e.g. `query_criteria_count`). This simply returns the number of associated data products instead of the full observation table. These are much faster, and are useful in project preparation.

In [19]:
query_obj_count = Observations.query_object_count('WASP-121b')
query_region_count = Observations.query_region_count('200.22, 10.33')
query_criteria_count = Observations.query_criteria_count(project = 'HST', dataproduct_type = "spectrum")

print(query_obj_count, query_region_count, query_criteria_count)

6446 74 211942


### Metadata Queries
Additionally, metadata can be queried to help users properly fill out the `query_criteria` method, and is an alternative to the list provided above.

In [20]:
#For observations
meta_data = Observations.get_metadata("observations")
print(meta_data[:2])

#For product types
meta_data = Observations.get_metadata("products")
print(meta_data[:5])

 Column Name     Column Label   ...       Examples/Valid Values       
-------------- ---------------- ... ----------------------------------
    intentType Observation Type ... Valid values: science, calibration
obs_collection          Mission ...          E.g. SWIFT, PS1, HST, IUE
  Column Name    ...
---------------- ...
          obs_id ...
           obsID ...
  obs_collection ...
dataproduct_type ...
     description ...


## Downloading Data Products

Data can be downloaded using the `get_product_list` and `download_products` methods. The general paradigm is:

- Generate an observation table as above
- From the observation table, generate a product list via `get_product_list`
- Download the data in the product list using `download_products`

This is shown below.

In [21]:
##Repeat our previous TESS query for HAT-P-7b
query_table = Observations.query_criteria(objectname="HAT-P-7b", 
                                              obs_collection="TESS",
                                              dataproduct_type=["timeseries"])
##Only the first two associated data sets
data_products = Observations.get_product_list(query_table[0:2])

##Download the data products in the designated folder
downloads = Observations.download_products(data_products, download_dir="Data")
print(downloads)

INFO: Found cached file Data/mastDownload/TESS/tess2019226182529-s0015-0000000424865046-0151-s/tess2019226182529-s0015-0000000424865046-0151-s_lc.fits with expected size 1906560. [astroquery.query]
INFO: Found cached file Data/mastDownload/TESS/tess2019226182529-s0015-0000000424865046-0151-s/tess2019226182529-s0015-0000000424865046-0151-s_tp.fits with expected size 45956160. [astroquery.query]
INFO: Found cached file Data/mastDownload/TESS/tess2019226182529-s0015-0000000424865156-0151-s/tess2019226182529-s0015-0000000424865156-0151-s_lc.fits with expected size 1906560. [astroquery.query]
INFO: Found cached file Data/mastDownload/TESS/tess2019226182529-s0015-0000000424865156-0151-s/tess2019226182529-s0015-0000000424865156-0151-s_tp.fits with expected size 45956160. [astroquery.query]
INFO: Found cached file Data/mastDownload/TESS/tess2019226182529-s0015-0000000424865156-0151-s/tess2019227203528-s0015-s0015-0000000424865156-00245_dvm.pdf with expected size 4889891. [astroquery.query]
INF

Additionally, you can download a single file by using the `download_file` method, which requires the product list to contain only one object.

In [22]:
#Go to the next product in our above list for HAT-P-7b
data_products = Observations.get_product_list(query_table[7:8])
##Select the first file
file = data_products["dataURI"][0]
##Check the extension of the retrieved file to match with the local path method below
print(file)

mast:TESS/product/tess2021205113501-s0041-s0041-0000000424865156-00511_dvm.pdf


In [23]:
## The local path needs to reference a file, not a folder in this case
import os
file_download = Observations.download_file(file, local_path=os.path.join('..', 'Data_Repositories', 'Data', 'TESS_Data.pdf'))

INFO: Found cached file ../Data_Repositories/Data/TESS_Data.pdf with expected size 5152165. [astroquery.query]


This gives a TESS SPOC report for the known planet HAT-P-7b, with a distinct transit. Nice work! If you want more tutorials or information on available methods, [here](https://astroquery.readthedocs.io/en/latest/mast/mast_obsquery.html#downloading-data) is the `astropy` documentation for downloading MAST products.

## Catalog Queries

Similar to the NASA Exoplanet Archive, MAST hosts a handful of catalogs that can be programmatically accessed via `astropy`. These include the:
- [Hubble Source Catalog](https://archive.stsci.edu/hst/hsc/) (HSC)
- [GALEX Catalog](http://www.galex.caltech.edu/researcher/data.html) (Galex)
- [Gaia and TGAS Catalogs](https://www.cosmos.esa.int/web/gaia/data-release-3) (Gaia)
- [TESS Input Catalog](https://tess.mit.edu/science/tess-input-catalogue/) (TIC)
- [TESS Candidate Target List](https://www.google.com/search?q=TESS+candidate+target+list&sca_esv=cc6cef90bdd03fb0&sca_upv=1&rlz=1C1CHBF_enUS910US910&sxsrf=ADLYWIJKOEN2t6r5zN2BYM77G3Iywl2kqw%3A1720631144906&ei=aL-OZsngNpDbwN4Pkde76AI&ved=0ahUKEwjJtLnB-pyHAxWQLdAFHZHrDi0Q4dUDCA8&uact=5&oq=TESS+candidate+target+list&gs_lp=Egxnd3Mtd2l6LXNlcnAiGlRFU1MgY2FuZGlkYXRlIHRhcmdldCBsaXN0MgUQIRigATIFECEYoAEyBRAhGKABSL0ZUABY6xhwAngBkAEAmAHyAaAB-BiqAQY1LjIxLjG4AQPIAQD4AQGYAh2gAsIZwgIKECMYgAQYJxiKBcICBBAjGCfCAgoQABiABBhDGIoFwgIKEC4YgAQYQxiKBcICERAuGIAEGLEDGNEDGIMBGMcBwgILEAAYgAQYsQMYgwHCAg4QLhiABBixAxjRAxjHAcICCBAuGIAEGLEDwgIQEAAYgAQYsQMYgwEYFBiHAsICCBAAGIAEGLEDwgILEC4YgAQYxwEYrwHCAg0QLhiABBixAxgUGIcCwgINEC4YgAQYFBjUAhiHAsICEBAuGIAEGLEDGEMYgwEYigXCAgsQLhiABBixAxiDAcICBRAAGIAEwgILEAAYgAQYkQIYigXCAgUQLhiABMICChAuGIAEGBQYhwLCAgsQLhiABBiRAhiKBcICChAuGIAEGLEDGArCAgoQABiABBgUGIcCwgIIEC4YgAQY1ALCAgcQLhiABBgKwgIGEAAYFhgewgIKEAAYFhgKGB4YD8ICCxAAGIAEGIYDGIoFwgIIEAAYgAQYogTCAgUQIRirApgDAJIHBjYuMjIuMaAH5JEC&sclient=gws-wiz-serp) (CTL)
- [Disk Directive Catalog]() (DiskDirective)
- [PanSTARRS Catalog](https://catalogs.mast.stsci.edu/docs/panstarrs.html) (PANSTARRS)
- [All-Sky PLATO Input Catalog](https://www.aanda.org/articles/aa/full_html/2021/09/aa40717-21/aa40717-21.html) (Plato)

Since both the NASA Exoplanet Archive and MAST catalog queries are both under `astropy`, the syntax is nearly identical. Reference the [NASA Exoplanet Archive module](..\Databases\NEA.ipynb) for how to construct `query_object`, `query_region`, and `query_criteria`. To query a specific catalog, use the `catalog = [catalog ID]` parameter with any of the `query` methods above, along with the identifier associated with the catalog, shown within the parentheses in the list above. Here is a quick example from the PLATO Input Catalog:

In [24]:
catalog_data = Catalogs.query_region("100.23 -10", catalog="Plato")
print(catalog_data)

 objID      PICnameDR1          sourceId      ... PICidDR1      distance     
-------- ---------------- ------------------- ... -------- ------------------
29307583 PIC DR1 29307583 3002281237690853504 ... 29307583 2.5791268100984066
29273768 PIC DR1 29273768 3002277080162573312 ... 29273768  3.412897833173899
29265435 PIC DR1 29265435 3002275705773086976 ... 29265435  6.003720626738184
29287541 PIC DR1 29287541 3002280035100117760 ... 29287541 7.9166526503407955
29306080 PIC DR1 29306080 3002328718551357440 ... 29306080  8.229280048747171
29301456 PIC DR1 29301456 3002327314099544192 ... 29301456 10.261798030413424
29359276 PIC DR1 29359276 3002424208563685376 ... 29359276 11.738937090205514
29235782 PIC DR1 29235782 3002268936905107072 ... 29235782 11.837705768101078


Note that not all query methods are available for each catalog. Look [here](https://astroquery.readthedocs.io/en/latest/mast/mast_catalog.html) for more examples, and [here](https://astroquery.readthedocs.io/en/latest/api/astroquery.mast.CatalogsClass.html#astroquery.mast.CatalogsClass) for documentation on the Catalog class.

# [Exoplanet Atmosphere Observability Table](https://catalogs.mast.stsci.edu/eaot/)
The last major tool that MAST offers specific to exoplanet research is the Exoplanet Atmosphere Observability Table (EAOT). This provides spectroscopic emission signal-to-noise ratios (SNRs) at 1.5 and 5 $\mathrm{\mu m}$, as well as the K-band transmission SNR. All of these values are **relative to [HD 209458 b](https://en.wikipedia.org/wiki/HD_209458_b)**. 

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_EAOT.PNG)

The results can be filtered based on a variety of different parameters. The output results look as follows:

![Logo ><](../../Auxiliary_Files/Graphics/Data_Repositories/MAST_EAOT_Results.PNG)

As always, the table can be exported to `.csv` or `.json`. 

# Exercises

## 1. Download TESS Light Curve Data
One of the primary data products from the TESS mission is time-series photometry. These are used to identify transits of orbiting exoplanets. In this exercise, you are tasked with querying and downloading TESS light curve file from the SPOC pipeline for TOI-1001.01. It has TIC ID 88863718, which will help you find the data you want.

HINT: You will want to query by criteria, using the `target_name`, `provenance_name`, and `dataproduct_type` parameters. `target_name` **must** be a string, even if the target name is an integer. `provenance_name` refers to the pipeline that produces the data, which is SPOC in this case. [Here](https://mast.stsci.edu/api/v0/_c_a_o_mfields.html) is the list of parameters for the Observation class. When querying the `product_list`, you will want to filter for `fits` extensions, and grab data that ends in `...s_lc.fits`.

In [25]:
##Codespace for Exercise 1

## 2. Previewing Hubble Data for Transmission Spectroscopy
Transmission spectroscopy utilizes atmospheric absorption of the exoplanet as it passes in front of its host star. To constrain chemical species in the atmosphere, Hubble's WFC3 infrared spectrometer can be used before and during transits, to see which wavelengths are absorbed. To do this, spectral images must be obtained. While this can be done via `astroquery` or the MAST portal, is it usually easier to access using the [Hubble Search Tool](https://mast.stsci.edu/search/ui/#/hst), as it is typically only a few dozen exposures with the right filters. Query for object KELT-20 b, filter G141, and observation time between 2023-02-20 and 2023-03-20. This will give you 119 spectra and 1 direct image of KELT-20 b, which you can preview and (optionally) download. If you download, select the spectra you'd like to analyze, select 'choose which files to download', and under 'calibrated,' choose 'IMA' files. This can be analyzed using the [PACMAN](..\Software\PACMAN.ipynb).

## 3. Using the NASA Exoplanet Archive with MAST to Systematically Access Exoplanet Host Data
The NASA Exoplanet Archive can easily give a list of known exoplanets with object identifiers that can be used to query e.g. TESS time-series data programmatically. This is done using the **TIC ID**. Following the [NASA Exoplanet Archive module](../Catalogs/NEA.ipynb), download a .csv file of all exoplanet hosts that:
- Have at least one transiting exoplanet
- Have $4000 \, \mathrm{K} \leq T_{eff} \leq 5000 \, \mathrm{K} $
- An exoplanet with radius $0.7 \, R_{\bigoplus} \leq R_{P} \leq 1.3 \, R_{\bigoplus}$
Then, create a Python script that systematically access the TIC IDs from the downloaded .csv, queries the data product **counts** of all TESS timeseries data, and stores it in a dictionary with the host name as the key. Which host has the greatest amount of TESS timeseries data associated with it?

In [26]:
##Codespace for Exercise 2